3p/ampcontext.js

```diff
@@ -0,0 +1,203 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import './polyfills';
+import {
+  dev,
+  user,
+} from '../src/log';
+import {IframeMessagingClient} from './iframe-messaging-client';
+
+/**
+   Enum for the different postmessage types for the window.context
+   postmess api.
+*/
+export const MessageType_ = {
+  SEND_EMBED_STATE: 'send-embed-state',
+  EMBED_STATE: 'embed-state',
+  SEND_EMBED_CONTEXT: 'send-embed-context',
+  EMBED_CONTEXT: 'embed-context',
+  SEND_INTERSECTIONS: 'send-intersections',
+  INTERSECTION: 'intersection',
+  EMBED_SIZE: 'embed-size',
+  EMBED_SIZE_CHANGED: 'embed-size-changed',
+  EMBED_SIZE_DENIED: 'embed-size-denied',
+};
+
+export class AmpContext extends IframeMessagingClient {
+
+  /**
+   *  @param {Window} win The window that the instance is built inside.
+   */
+  constructor(win) {
+    super(win);
+    this.setupMetadata_();
+    this.ampWindow = this.getHostWindow();
+  }
+
+  /** @override */
+  getHostWindow() {
```

Comment:
getHostWindow as a public method, can potentially be called multiple times, should return a cache when available.

we can move the computation logic into constructor.
or, if you want to do lazy initialization, which I  don't see much value here, you can initialize this.ampWindow here. Then all the other code should not  reference this.ampWindow directly, they should reference this method.